cardano-db/src/Cardano/Db/Operations/Other/ConsumedTxOut.hs

```diff
@@ -0,0 +1,541 @@
+updatePageEntries ::
+  MonadIO m =>
+  TxOutTableType ->
+  [ConsumedTriplet] ->
+  ReaderT SqlBackend m ()
+updatePageEntries txOutTableType = mapM_ (updateTxOutConsumedByTxIdUnique txOutTableType)
+
+updateTxOutConsumedByTxIdUnique :: MonadIO m => TxOutTableType -> ConsumedTriplet -> ReaderT SqlBackend m ()
+updateTxOutConsumedByTxIdUnique txOutTableType ConsumedTriplet {ctTxOutTxId, ctTxOutIndex, ctTxInTxId} =
+  case txOutTableType of
+    TxOutCore -> updateWhere [C.TxOutTxId ==. ctTxOutTxId, C.TxOutIndex ==. ctTxOutIndex] [C.TxOutConsumedByTxId =. Just ctTxInTxId]
+    TxOutVariantAddress -> updateWhere [V.TxOutTxId ==. ctTxOutTxId, V.TxOutIndex ==. ctTxOutIndex] [V.TxOutConsumedByTxId =. Just ctTxInTxId]
+
+-- this builds up a single delete query using the pageEntries list
+deletePageEntries ::
+  MonadIO m =>
+  TxOutTableType ->
+  [ConsumedTriplet] ->
+  ReaderT SqlBackend m ()
+deletePageEntries txOutTableType = mapM_ (\ConsumedTriplet {ctTxOutTxId, ctTxOutIndex} -> deleteTxOutConsumed txOutTableType ctTxOutTxId ctTxOutIndex)
+
+deleteTxOutConsumed :: MonadIO m => TxOutTableType -> TxId -> Word64 -> ReaderT SqlBackend m ()
+deleteTxOutConsumed txOutTableType txOutId index = case txOutTableType of
+  TxOutCore -> deleteWhere [C.TxOutTxId ==. txOutId, C.TxOutIndex ==. index]
+  TxOutVariantAddress -> deleteWhere [V.TxOutTxId ==. txOutId, V.TxOutIndex ==. index]
+
+--------------------------------------------------------------------------------------------------
+-- Raw Queries
+--------------------------------------------------------------------------------------------------
+createConsumedTxOut ::
+  forall m.
+  ( MonadBaseControl IO m
+  , MonadIO m
+  ) =>
+  ReaderT SqlBackend m ()
+createConsumedTxOut = do
+  handle exceptHandler $ rawExecute createIndex []
+  handle exceptHandler $ rawExecute addConstraint []
+  where
+    createIndex =
+      "CREATE INDEX IF NOT EXISTS idx_tx_out_consumed_by_tx_id ON tx_out (consumed_by_tx_id)"
+
+    addConstraint =
+      ( Text.unlines
+          [ "do $$"
+          , "begin"
+          , "  if not exists ("
+          , "    select 1"
+          , "    from information_schema.table_constraints"
+          , "    where constraint_name = 'ma_tx_out_tx_out_id_fkey'"
+          , "      and table_name = 'ma_tx_out'"
+          , "  ) then"
+          , "    execute 'alter table ma_tx_out add constraint ma_tx_out_tx_out_id_fkey foreign key(tx_out_id) references tx_out(id) on delete cascade on update restrict';"
+          , "  end if;"
+          , "end $$;"
+          ]
+      )
+
+    exceptHandler :: SqlError -> ReaderT SqlBackend m a
+    exceptHandler e =
+      liftIO $ throwIO (DBPruneConsumed $ show e)
+
+updateTxOutAndCreateAddress ::
+  forall m.
+  ( MonadBaseControl IO m
+  , MonadIO m
+  ) =>
+  ReaderT SqlBackend m ()
+updateTxOutAndCreateAddress = do
+  handle exceptHandler $ rawExecute alterTxOutQuery []
+  handle exceptHandler $ rawExecute createAddressTableQuery []
```

Comment:
yeah looks like I'll have to also remove existing indexes/views too:
```
DBExtraMigrationSqlError 
  { sqlState = "2BP01", 
    sqlExecStatus = FatalError, 
    sqlErrorMsg = "cannot drop column address of table tx_out because other objects depend on it", 
    sqlErrorDetail = "view utxo_byron_view depends on column address of table tx_out\nview utxo_view depends on column address of table tx_out", 
    sqlErrorHint = "Use DROP ... CASCADE to drop the dependent objects too."
  }
 ```